sock: initial definitions for asynchronous event handling

miri64 opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description
This represents another approach to asynchronous sock as outlined in https://github.com/RIOT-OS/RIOT/pull/8149#issuecomment-482497589. Apart from the necessary definitions for all basic sock types, this PR also contains example implementations for `event.h` and `msg.h` which can also be separated out into another PR.
<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure
No implementation in this PR, just API (and a implementations using this unimplemented API ;-)).
<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references
Alternative to #8149.
<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

kaspar030 commented:
Nice move forward!

Some thoughts (only brief look):

- if we introduce a generic ```sock_t``` first, that all other sock types (udp, tcp, ip, ...) inherit from, the API gets a lot smaller, and the async handling part can be written once, for all of them. And IMO, as now we're starting to get shared funtionality, this would actually make sense to introduce.
I'm envisioning only a couple of ops, like read(), write(), close(), status(), setup_async_event() .

- the event type needs to be OR'ed.

If I see correctly, there's one "ctx" per sock, and thus one event object. If now there are multiple events (CAN_READ | CLOSED), the second that arrives overwrites the first. The event should be used as bitfield, and it should be set in a way that even if the event is already queued, a subsequent bit that gets set can still be read.


- Do you think we need to go all generic with the callback method?

I think we might get away with just ifdef'ing support for the different event methods, without resorting to custom types.
Like, only ```#ifdef MODULE_SOCK_ASYNC_EVENT```, the necessary fields are added to "sock_x_t", and the helper functions are included.

miri64 commented:
> * if we introduce a generic `sock_t` first, that all other sock types (udp, tcp, ip, ...) inherit from, the API gets a lot smaller, and the async handling part can be written once, for all of them. And IMO, as now we're starting to get shared funtionality, this would actually make sense to introduce.
>   I'm envisioning only a couple of ops, like read(), write(), close(), status(), setup_async_event() .

I was thinking about this, [actually even had some header definitions for that](https://github.com/miri64/RIOT/commit/6bac5ce9181fa534d2f29df6c25cd234996139b8) however those were transparent to the lower levels of the implementation and thus did not really simplify things. Everything beyond that would rather be a completely new API, which would not `sock` and thus has nothing to do with my proposal.

> * the event type needs to be OR'ed.

They will be. By the lower level implementation. As you can see for `sock_event_type_t`, the values are OR'able ;-).

> If I see correctly, there's one "ctx" per sock, and thus one event object.

`ctx` is very implementation specific, as you may have noticed, so it could also be a queue of objects.

> If now there are multiple events (CAN_READ | CLOSED), the second that arrives overwrites the first. The event should be used as bitfield, and it should be set in a way that even if the event is already queued, a subsequent bit that gets set can still be read.

There are ways to prevent that (e.g. making sure the callback is only once and then again after an event was handled)

> * Do you think we need to go all generic with the callback method?

As pointed out above: I'm against changing the current API (apart from amendments and minor changes) at this point.

> I think we might get away with just ifdef'ing support for the different event methods, without resorting to custom types.
> Like, only `#ifdef MODULE_SOCK_ASYNC_EVENT`, the necessary fields are added to "sock_x_t", and the helper functions are included.

That is the idea of `SOCK_HAS_ASYNC`

miri64 commented:
Another reason against using a "generic sock" here would be that it then wouldn't be applicable for TCP listener queues.

miri64 commented:
I added another example implementation using `msg_t` which could come in handy at some point for fixing https://github.com/RIOT-OS/RIOT/issues/8130

haukepetersen commented:
Very nice, can't wait to have axync sock!

I am not 100% sure if I understood the code correctly, but I don't like the direct mapping to `event` and `msg`. I would think that it is more flexible and future proof, if we would introduce a classical callback interface without any dependency to any IPC implemenation as base. 

We can then use this 'lower-level' to implement the usual convenience wrappers for concrete IPC implementations (`msg`, `event`, `thread_flags`) on top of that, pretty much like the `xtimer` interface. This way we can easily map any IPC that might pop up in the future, but also users are able to decide on their own way to handle sock events... Also it makes it easy to select the needed functionality, as in 'USEMODULE sock_async_thread_flags` etc...

If this is already the case and I just did not read the code of this PR correctly then never mind :-)

miri64 commented:
> I am not 100% sure if I understood the code correctly, but I don't like the direct mapping to `event` and `msg`. I would think that it is more flexible and future proof, if we would introduce a classical callback interface without any dependency to any IPC implemenation as base.

There is no direct mapping to `event` and `msg`. The code mapping to `event` and `msg` are just example implementations of how the map the callback structure of async sock to RIOT... So what you are proposing is already the case here.

miri64 commented:
See e.g. https://github.com/RIOT-OS/RIOT/pull/11723/commits/8d33aed73e503bf1556972ca815e2f1338b0432f for how async sock is handled for `sock_udp`. If you want some thing more complicated and abstract than just callbacks, we can go back to https://github.com/RIOT-OS/RIOT/pull/8149 but that was stalled for years...

miri64 commented:
> If this is already the case and I just did not read the code of this PR correctly then never mind :-)

Sorry, should have read your comment to the very end :sweat_smile:. Yes, what you are proposing is already the case.

haukepetersen commented:
Now I saw, I mainly missed the difference of `sock_event_type_t` and `sock_event_t` :-)

miri64 commented:
I gave this PR some well deserved love:

- Rebased to current master
- Added DTLS support
- Made sure the `sock_async_*` modules are compilable ;-)

All that's missing is some dock on the `sock_async_*` modules IMHO and then it might be ready to go (of course the network stacks need to implement the event triggering as well, but that I rather would do as a follow-up).

miri64 commented:
Added some doc for `sock_async_event`.

miri64 commented:
OK. I updated all the doc now. Anybody wants to review. I did not test, if the examples I provide work, as there is no implementation of this API yet, but I'd be happy to provide one for GNRC and lwIP, once I get some positive feedback.

miri64 commented:
If people prefer it, I can factor out the example front-end implementations into separate PRs.

miri64 commented:
@haukepetersen @kaspar030 et al. can I please have some feedback on this?

kaspar030 commented on sys/net/sock/async/event/sock_async_event.c:
```diff
@@ -0,0 +1,103 @@
+    ctx->event.sock = sock;
+    ctx->event.type |= type;
+    event_post(ctx->queue, &ctx->event.super);
+    ctx->event.type &= ~type;
```
this must stay set until after the cb has been executed.

miri64 commented:
Squashed first (and renamed the event type flags as they caused some confusion offline)

miri64 commented:
And squashed and renamed again (forgot to rename the usage of that type)

miri64 commented:
Aand removed the implementations. Ready for review again.

kaspar030 commented on sys/include/net/sock.h:
```diff
@@ -153,6 +163,23 @@ extern "C" {
+ * @note    Only applicable with @ref SOCK_HAS_ASYNC defined.
+ */
+typedef enum {
+    SOCK_EVENT_CONN_RDY  = 0x0001,  /**< Connection ready event */
```
maybe `SOCK_ASYNC_foo`?

kaspar030 commented on sys/include/net/sock/dtls.h:
```diff
@@ -636,6 +656,41 @@ ssize_t sock_dtls_send(sock_dtls_t *sock, sock_dtls_session_t *remote,
+
+#if SOCK_HAS_ASYNC_CTX || defined(DOXYGEN)
+/**
+ * @brief   Gets the asynchronous
```
.

miri64 commented:
> We discussed offline that while this API is not stable we'll keep it separate from the existing sock headers and just copy all changes from this PR into a shared `sock/async.h`.
> That way existing code, API and headers stay untouched for now.

Done and squashed (as I removed most commits anyway).
